server/src/main/java/keywhiz/service/config/ClientAuthConfig.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package keywhiz.service.config;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import java.util.List;
+
+/**
+ * Configuration for how clients should be authenticated.
```

Comment:
Having reviewed the PR, I'm still unclear as to how xfcccConfigs and typeConfig are related. xfccConfigs appears to declare which ports and mTLS clients to trust for providing client identity via XFCC, whereas typeConfig appears to be about how to extract client identity from a client certificate. What's still confusing to me is whether typeConfig rules will apply to client certificates provided via XFCC. It looks like they don't apply, but I'm not sure why they shouldn't.

Reply:
That's a good summary of the two configs. 

Currently, typeConfig applies to retrieving a Keywhiz client entity based on either the mTLS certificate (for a non-XFCC port) or the certificate parsed from the XFCC header (for an XFCC port). I'm not sure if the second case is what you meant by "client certificates provided via XFCC."

It doesn't apply to the entity parsed from the mTLS certificate for an XFCC port, which the code calls the "requester" rather than the "client" since it's probably a proxy. That does mean that a requester/proxy that sets both the CN and the SPIFFE ID on its certificate needs to be listed in _both_ allowlists. It might be worth allowing the configuration to explicitly disable an allowlist (although that makes ensuring that all traffic is checked against at least one allowlist slightly more complex).

I'd still use a different configuration to control disabling an allowlist, though, to draw a distinction between the typeConfig that controls how the Keywhiz client objects are identified (by name or by spiffe ID) and the allowlists that control whether XFCC headers are permitted. (I probably should have postponed the typeConfig code to a later PR.)